node/src/components/consensus/protocols/highway/performance_meter.rs

```diff
@@ -0,0 +1,147 @@
+use std::cmp::{max, min};
+
+use datasize::DataSize;
+use tracing::error;
+
+use casper_types::{TimeDiff, Timestamp};
+
+use crate::components::consensus::{
+    highway_core::{
+        finality_detector::{
+            assigned_weight_and_latest_unit, find_max_quora, round_participation,
+            RoundParticipation,
+        },
+        state, State,
+    },
+    traits::Context,
+    utils::ValidatorIndex,
+};
+
+/// The number of most recent blocks for which we average the max quorum in which we participated.
+const BLOCKS_TO_CONSIDER: usize = 5;
+/// The average max quorum that triggers us to slow down: with this big or smaller average max
+/// quorum per `BLOCKS_TO_CONSIDER`, we increase our round length.
+const SLOW_DOWN_THRESHOLD: f64 = 0.8;
+/// The average max quorum that triggers us to speed up: with this big or larger average max quorum
+/// per `BLOCKS_TO_CONSIDER`, we decrease our round length.
+const ACCELERATION_THRESHOLD: f64 = 0.9;
+/// We will try to accelerate (decrease our round length) every `ACCELERATION_PARAMETER` rounds if
+/// we have a big enough average max quorum.
+const ACCELERATION_PARAMETER: u64 = 40;
+
+#[derive(DataSize, Debug, Clone)]
+pub(crate) struct PerformanceMeter {
+    our_vid: ValidatorIndex,
+    min_round_len: TimeDiff,
+    max_round_len: TimeDiff,
+    current_round_len: TimeDiff,
+    last_switch_round_id: Timestamp,
+}
+
+impl PerformanceMeter {
+    pub fn new(
+        our_vid: ValidatorIndex,
+        round_len: TimeDiff,
+        min_round_len: TimeDiff,
+        max_round_len: TimeDiff,
+        timestamp: Timestamp,
+    ) -> Self {
+        let current_round_id = state::round_id(timestamp, round_len);
+        Self {
+            our_vid,
+            min_round_len,
+            max_round_len,
+            current_round_len: round_len,
+            last_switch_round_id: current_round_id,
+        }
+    }
+
+    pub fn current_round_len(&self) -> TimeDiff {
+        self.current_round_len
+    }
+
+    /// If the current timestamp indicates that the round has ended, checks the known proposals for
+    /// a level-1 summit.
+    /// If there is a summit, the round is considered successful. Otherwise, it is considered
+    /// failed.
+    /// Next, a number of last rounds are being checked for success and if not enough of them are
+    /// successful, we return a higher round length for the future.
+    /// If the length shouldn't grow, and the round ID is divisible by a certain number, a lower
+    /// round length is returned.
```

Comment:
this comment is out of date; can you update the comment to describe the new functionality?